File: c++/Neetcode_150/3_Stacks/16_evaluate_rpe.cpp
//Evaluate Reverse Polish Expression: RPE => array contating nos. and operands are give and each operand applies to last 2 nos. on left of the operand. Evaluate the array and return the ans val.

int evalRPN(vector<string>& tokens) {
        stack<int> st;
        unordered_map<string, function<int(int,int)>> operators = {
            {"+", [](int a, int b) { return a+b; }},
            {"-", [](int a, int b) { return a-b; }},
            {"*", [](int a, int b) { return a*b; }},
            {"/", [](int a, int b) { return a/b; }},
        };
        for(auto& t: tokens){
            if(operators.find(t) != operators.end()){
                int a = st.top();
                st.pop();
                int b = st.top();
                st.pop();
                st.push(operators[t](b, a));
            }
            else st.push(stoi(t));
        }
        return st.top();
    }

File: js/Neetcode_150/3_Stacks/16_evaluate_rpe.js
//Evaluate Reverse Polish Expression: RPE => array contating nos. and operands are give and each operand applies to last 2 nos. on left of the operand. Evaluate the array and return the ans val.

//S1: Short Solution 

let stack = []
    const operators = {  //mapped operators with fxns. that perform the operation & give result
        '+': (b, a) => a + b,
        '-': (b, a) => a - b,
        '*': (b, a) => a * b,
        '/': (b, a) => Math.trunc(a / b)
    };
    for(let t of tokens){
        if(operators[t]){
            stack.push(operators[t](stack.pop(), stack.pop()))
        }
        else stack.push(Number(t))
    }
    return stack[stack.length-1]
};

// S2: Long Solution 

var evalRPN = function(tokens) {
    let stack = []
    for(let t of tokens){
        if(t === '+'){
            let a = stack.pop()
            let b = stack.pop()
            stack.push(a+b)
        }
        else if(t === '*'){
            let a = stack.pop()
            let b = stack.pop()
            stack.push(a*b)
        }
        else if(t === '-'){
            let a = stack.pop()
            let b = stack.pop()
            stack.push(b-a)
        }
        else if(t === '/'){
            let a = stack.pop()
            let b = stack.pop()
            stack.push(Math.trunc(b/a))
        }
        else stack.push(Number(t))
    }
    return stack[stack.length-1]
};